transport/httptransport/libp2p_server.go

```diff
@@ -225,15 +228,49 @@ func (s *Libp2pCarServer) sendCar(r *http.Request, w http.ResponseWriter, val *A
 
 	// Check if there was an error during the transfer
 	if err != nil {
-		log.Infow("transfer failed", append(logParams, "error", err))
+		log.Infow("transfer failed", append(logParams, "err", err)...)
+
+		st := xfer.setComplete(err)
+		fireEvent(st)
+		return err
+	}
+
+	// Wait for the client to receive all data and close the connection
+	err = waitForClientClose(s.ctx, r)
+	if err == nil {
+		log.Infow("completed serving request", logParams...)
 	} else {
-		log.Infow("completed serving request", logParams)
+		log.Infow("error waiting for client to close connection", append(logParams, "err", err)...)
 	}
 
 	st := xfer.setComplete(err)
 	fireEvent(st)
 
-	return nil
+	return err
+}
+
+// waitForClientClose waits for the client to close the connection
+func waitForClientClose(ctx context.Context, r *http.Request) error {
+	streamClosed := make(chan error, 1)
+	go func() {
+		// Block until Read returns an EOF, which means the connection has
+		// been closed
+		_, err := r.Body.Read(make([]byte, 1024))
```

Comment:
What if the server simply dosen't write anything ? Does this read call request the request context so we don;t leak these go routines ?

Reply:
Discussed sync: when the http handler returns, it will close the stream, so the Read call with return with an io.EOF